tools/lint/src/main/kotlin/NonAndroidxNullabilityDetector.kt

```diff
@@ -0,0 +1,131 @@
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.LintFix
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.isKotlin
+import org.jetbrains.uast.UAnnotation
+import org.jetbrains.uast.UClass
+import org.jetbrains.uast.UDeclaration
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.UMethod
+
+private val NULLABLE_ANNOTATIONS = listOf("Nullable", "CheckForNull")
+private val NOT_NULL_ANNOTATIONS = listOf("NonNull", "NotNull", "Nonnull")
+
+class NonAndroidxNullabilityDetector : Detector(), SourceCodeScanner {
+    companion object Issues {
+        private val IMPLEMENTATION = Implementation(
+                NonAndroidxNullabilityDetector::class.java,
+                Scope.JAVA_FILE_SCOPE
+        )
+
+        @JvmField
+        val NON_ANDROIDX_NULLABILITY = Issue.create(
+                id = "FirebaseNonAndroidxNullability",
+                briefDescription = "Use androidx nullability annotations.",
+
+                explanation = "Use androidx nullability annotations instead.",
+                category = Category.COMPLIANCE,
+                priority = 1,
+                severity = Severity.ERROR,
+                enabledByDefault = false,
+                implementation = IMPLEMENTATION
+        )
+    }
+
+    override fun getApplicableUastTypes(): List<Class<out UElement>>? {
+        return listOf(UClass::class.java, UMethod::class.java, UField::class.java)
+    }
+
+    override fun createUastHandler(context: JavaContext): UElementHandler? {
+        // using deprecated psi field here instead of sourcePsi because the IDE
+        // still uses older version of UAST
+        if (isKotlin(context.uastFile?.sourcePsi)) {
+            // These checks apply only to Java code
+            return null
+        }
+        return Visitor(context)
+    }
+
+    class Visitor(private val context: JavaContext) : UElementHandler() {
+        override fun visitClass(node: UClass) {
+            doVisit(node)
+        }
+
+        override fun visitMethod(node: UMethod) {
+            doVisit(node)
+            for (parameter in node.uastParameters) {
+                doVisit(parameter)
+            }
+        }
+
+        override fun visitField(node: UField) {
+            doVisit(node)
+        }
+
+        private fun doVisit(node: UDeclaration) {
+            for (annotation in node.annotations) {
+                if (isNullabilityAnnotation(annotation)) {
+                    ensureAndroidNullability(context, annotation)
+                }
+            }
+        }
+
+        private fun ensureAndroidNullability(context: JavaContext, annotation: UAnnotation) {
+            annotation.qualifiedName?.let { name ->
```

Comment:
Though less fancy you might consider returning early if qualifiedName is null to save indentation.

It's not really a problem here since it's just the one condition, but it makes things much easier to understand in the presence of multiple. See also http://go/tott-455.

Reply:
good point done